### PR TITLE
Contributors guide version 1

### PR DESCRIPTION
**Pull request by worekleszczy**
## Summary:

Draft to agree on contributors guide

### CODE REVIEW

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+
+
+# How to contribute
+
+Mesmer is a project to extract and export akka specific metrics.
+
+Projects is comprised of 3 major parts:
+- java agent (called `agent` from now on) - a piece of technology allowing to modify classes during load.
+- akka extension (called `extension` from now on)  - this is where all custom metrics computation happens.
+- application that is being instrumented - in repo example, but this in theory this could be any application that uses akka.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:10]:
```suggestion
- application that is being instrumented - in repo we have the "example" app, but this in theory this could be any application that uses akka.
```

File: CONTRIBUTORS.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+The most interesting parts are java agent and akka extension. They have to work in tandem - `agent` creates a foundation on which `extension` can build and calculate metrics.
+It's good to use an example to get a better understanding - `agent` allows to trigger an event whenever http request starts and completed
```
Comment (lgajowy) [inline CONTRIBUTORS.md:13]:
```suggestion
It's good to use an example to get a better understanding - `agent` allows to trigger an event whenever http request starts and completes
```

File: CONTRIBUTORS.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+Mesmer is a project to extract and export akka specific metrics.
+
+Projects is comprised of 3 major parts:
+- java agent (called `agent` from now on) - a piece of technology allowing to modify classes during load.
+- akka extension (called `extension` from now on)  - this is where all custom metrics computation happens.
+- application that is being instrumented - in repo example, but this in theory this could be any application that uses akka.
+
+The most interesting parts are java agent and akka extension. They have to work in tandem - `agent` creates a foundation on which `extension` can build and calculate metrics.
+It's good to use an example to get a better understanding - `agent` allows to trigger an event whenever http request starts and completed 
+(we found places where those should be triggered by looking at akka source code and then modify those classes during load).
```
Comment (lgajowy) [inline CONTRIBUTORS.md:14]:
```suggestion
We've found places where those events should be triggered by looking at Akka source code and then modified necessary classes during load). 
```

File: CONTRIBUTORS.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Projects is comprised of 3 major parts:
+- java agent (called `agent` from now on) - a piece of technology allowing to modify classes during load.
+- akka extension (called `extension` from now on)  - this is where all custom metrics computation happens.
+- application that is being instrumented - in repo example, but this in theory this could be any application that uses akka.
+
+The most interesting parts are java agent and akka extension. They have to work in tandem - `agent` creates a foundation on which `extension` can build and calculate metrics.
+It's good to use an example to get a better understanding - `agent` allows to trigger an event whenever http request starts and completed 
+(we found places where those should be triggered by looking at akka source code and then modify those classes during load). 
+`extension` then catch those events and calculate duration between the two and save it in opentelemetry instrument. There will be more about opentelemetry role in the product later in this document.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:15]:
```suggestion
`extension` then captures those events, calculates the duration between the two, and passes this calculated metric (request time) in opentelemetry instrument. There will be more about opentelemetry role in the product later in this document.
```

File: CONTRIBUTORS.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+## Extension
+
+It's a part that is responsible for calculating and exporting metrics. The fact that it's akka extension is an implementation detail, but it's worth to know that.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:20]:
```suggestion
It's a part that is responsible for calculating and exporting metrics. The fact that it's an Akka extension is an implementation detail, but it's worth knowing that.
```

File: CONTRIBUTORS.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+Akka extension are plugins that grant additional functionality to akka and can be enabled with configuration - this is the approach that we took.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:21]:
```suggestion
Akka extensions are plugins that grant additional functionality to Akka and can be enabled with configuration - this is the approach that we took.
```

It might make sense to replace this with a link to Akka extension docs for the users to get the general idea.

File: CONTRIBUTORS.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+At the start up of the extension we check for configured metrics that were enabled and start only those monitors that are required.
+In our example monitors are akka system actors, but this is also an implementation detail.
+As an example `io.scalac.mesmer.extension.HttpEventsActor` is a monitor that listens for AkkaHttp events.
+For conveyor of messages we use Reception
+
+
+## Opentelemetry
+
+It's a fusion of opentracing and openCensus.  It's an specification and an implementation for creating, aggregating and publishing traces and metrics.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:30]:
```suggestion
It's a fusion of opentracing and openCensus. It's an specification and an implementation for creating, aggregating and publishing traces and metrics.
```

File: CONTRIBUTORS.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+We use this extensively as it's makes publishing metrics to different backends (prometheus, datadog etc) easy and can take care of some aggregations (mostly histograms ATM).
+Links to read more:
+- https://opentelemetry.io/docs/instrumentation/java/
+- https://github.com/open-telemetry/opentelemetry-specification
+- https://github.com/open-telemetry/opentelemetry-java
+
+
+## Agent
+
+This is probably the most complicated part of mesmer development  (there is an ongoing proposal to switch from our custom agent to opentelemetry one which would  simplify development).
```
Comment (lgajowy) [inline CONTRIBUTORS.md:40]:
```suggestion
This is probably the most complicated part of mesmer development (there is an ongoing [proposal](https://github.com/ScalaConsultants/mesmer-akka-agent/discussions/272) to switch from our custom agent to opentelemetry one which would simplify development and enable tracing).
```

File: CONTRIBUTORS.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+In some cases it requires writing `.java` files as scala doesn't allow some operations that are needed  for class instrumentations.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:41]:
```suggestion
In some cases it requires writing `.java` files as scala doesn't allow some operations that are needed for class instrumentations.
```
I'd remove that sentence here and adress it in a "how to add new metrics" section.

File: CONTRIBUTORS.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+We use bytebyddy as a preferred tool for bytecode generation - it operates on higher level that ASM and has an `Advice` abstraction - this can be thing as a blueprint that will be injected into loaded class file.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:42]:
```suggestion
We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode generation - it operates on a higher level than ASM and has an `Advice` abstraction - this can be thing as a blueprint that will be injected into loaded class file.
```

Comment (lgajowy) [reply CONTRIBUTORS.md:42]: > this can be thing as a blueprint that will be injected into loaded class file. 

I don't understand this part - could you rephrase it?

File: CONTRIBUTORS.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+Advice api is mostly annotation driven.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:43]:
```suggestion
```

That's a detail imho :)

File: CONTRIBUTORS.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+Links to read more:
+- https://bytebuddy.net/#/tutorial
+- https://www.youtube.com/watch?v=oflzFGONG08 - introduction to bytebuddy and advices by it's creator Rafael Winterhalter
```
Comment (lgajowy) [inline CONTRIBUTORS.md:47]: Can you add a "how to add a new metric section"  below that will contain steps to do that? Even I could benefit from that (still not 100% sure what the steps are).

**Review by lgajowy — CHANGES_REQUESTED** [review]
Thanks! I think that the only think the doc is missing is a small guide on how to add new metrics. Could be done in this doc or in a separate one. If you have a good PR in mind that was adding a metric in the past, feel free to refer it in the doc as well.

File: CONTRIBUTORS.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+Links to read more:
+- https://opentelemetry.io/docs/instrumentation/java/
+- https://github.com/open-telemetry/opentelemetry-specification
+- https://github.com/open-telemetry/opentelemetry-java
+
+
+## Agent
+
+This is probably the most complicated part of mesmer development (there is an ongoing [proposal](https://github.com/ScalaConsultants/mesmer-akka-agent/discussions/272) to switch from our custom agent to opentelemetry one which would simplify development and enable tracing).
+In some cases it requires writing `.java` files as scala doesn't allow some operations that are needed  for class instrumentations.
```
Comment (worekleszczy) [inline CONTRIBUTORS.md]:
```suggestion

```

File: CONTRIBUTORS.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+- https://opentelemetry.io/docs/instrumentation/java/
+- https://github.com/open-telemetry/opentelemetry-specification
+- https://github.com/open-telemetry/opentelemetry-java
+
+
+## Agent
+
+This is probably the most complicated part of mesmer development (there is an ongoing [proposal](https://github.com/ScalaConsultants/mesmer-akka-agent/discussions/272) to switch from our custom agent to opentelemetry one which would simplify development and enable tracing).
+In some cases it requires writing `.java` files as scala doesn't allow some operations that are needed for class instrumentations.
+We use bytebyddy as a preferred tool for bytecode generation - it operates on higher level that ASM and has an `Advice` abstraction - this can be thing as a blueprint that will be injected into loaded class file.
```
Comment (worekleszczy) [inline CONTRIBUTORS.md]:
```suggestion
We use bytebyddy as a preferred tool for bytecode generation - it operates on higher level that ASM and has an `Advice` abstraction - it can be thought as a template that will be applied to a class file.
```

File: CONTRIBUTORS.md (hunk 10 above)
Comment (worekleszczy) [reply CONTRIBUTORS.md]: Yeah, check it now 🙂

File: CONTRIBUTORS.md (hunk 12 above)
Comment (worekleszczy) [reply CONTRIBUTORS.md:46]: Yeah, I must admit I didn't really follow a well thought process - but if I were to summarise is it would be a couple of simple steps:
* take a metric from the matrix
* check in akka source code how it can be implemented - this it the most tricky part, as we don't really know if something can be achieved and in which way. It's worth to know how akka works internally but this part is mostly trail and error. Sometimes, especially in "more modern" parts of akka code like `akka-persistence` or `akka-cluster` there is `@InternalStableApi` on methods that are stubs and probably overriden by lightbend monitoring agent.
* add exporting and if needed calculations in `extension`

Comment (worekleszczy) [reply CONTRIBUTORS.md:46]: But I'll think if I can add something here as I see that this is not a huge help

**Comment by adamgadomski** [comment]
resolves #302

File: CONTRIBUTORS.md (hunk 12 above)
Comment (lgajowy) [reply CONTRIBUTORS.md:46]: Something along these lines would be valuable indeed. For example: 

1. choose an API that you want to instrument with metrics
2. Make sure that you are able to test this instrumentation - if it's an Akka library, see if our example app uses that. If not, create a simple example app for testing that metric (consider contributing it to the "examples" directory in Mesmer.
3. Check how the instrumentation can be instrumented. What class should be modified by Bytebuddy? What method should report the metrics. What are the metrics, what is their kind (counter, gauge, histogram)?

I would also add something about: 

4. What is Agent* class? How to add it? How to enable it for the agent? (Show an example of an agent class)
5. What is event bus? What is it's purpose?
6. What is an <Metric>Actor class and what does it do? Where's the place where opentelemetry metrics API is actually used?

4, 5, 6 will possibly simplify in the future but for now it's relevant.

File: CONTRIBUTORS.md [hunk 15]
```diff
@@ -43,4 +43,30 @@ We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode ge
...
\ No newline at end of file
+- https://www.youtube.com/watch?v=oflzFGONG08 - introduction to bytebuddy and advices by it's creator Rafael Winterhalter
+
+## How to add a new metric
+
+Although adding a new metrics might be a little tricky, there are a couple of steps that are worth following:
+
+* choose a metric that you want to instrument
+* check which api might expose this metric
+* start with tests first - check if you have an application that use chosen API e.g. if if it's an Akka library, see if our example app uses that.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:54]:
```suggestion
* start with tests first - check if you have an application that use chosen API e.g. if it's an Akka library, see if our example app uses that.
```

File: CONTRIBUTORS.md [hunk 16]
```diff
@@ -43,4 +43,30 @@ We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode ge
...
+If not, create a simple example app for testing (consider contributing it to the `examples` directory in Mesmer)
+* choose what is a kind of metric e.g. is a gauge, counter or histogram?
```
Comment (lgajowy) [inline CONTRIBUTORS.md:56]:
```suggestion
* choose what is the kind of metric e.g. is a gauge, counter or histogram?
```

File: CONTRIBUTORS.md [hunk 17]
```diff
@@ -43,4 +43,30 @@ We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode ge
...
+* dive into source library source code - find a fragments of code that you can hook on to extract a metric.
```
Comment (lgajowy) [inline CONTRIBUTORS.md:57]:
```suggestion
* dive into library source code - find fragments of code that you can hook on to extract a metric.
```

File: CONTRIBUTORS.md [hunk 18]
```diff
@@ -43,4 +43,30 @@ We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode ge
...
+This is a tricky part and in many cases you might be forced to change previously made decision (e.g. change the metric type).
+This might get overwhelming quickly, so try to deliver a working PoC rather than figuring out all the details upfront.
+* write `agent` instrumentation code - this varies from metric to metric, but there are a couple rules of thumb
+  * don't use mutable state - currently preferred approach is to fire an event with `EventBus`
+  * use `Advice` rather than `Intercept` API
+  * use java classes as last resort and try to make it as thin as possible - it's usually better to write logic in scala functions
+* write monitor classes - current approach is not to use `opentelemetry` monitors directly but to use our `wrappers` which allowed for better testability.
+Check `io.scalac.mesmer.extension.metric` package in `extension` for current implementations.
+* write events handling class - currently those are solely actors, but this might change in the future. This is the part that should take
```
Comment (lgajowy) [inline CONTRIBUTORS.md:66]:
```suggestion
* write an event handling class - currently those are solely actors, but this might change in the future. This is the part that should take
```

File: CONTRIBUTORS.md [hunk 19]
```diff
@@ -43,4 +43,30 @@ We use [ByteBuddy](https://bytebuddy.net/#/) as a preferred tool for bytecode ge
...
+previously defined monitor as a parameter and submit values to it after it's done doing calculations.
+* remember about tests for both `extension` and `agent` part
+
+### Writing agent
+
+// TODO here will be content for adding agents
```
Comment (lgajowy) [inline CONTRIBUTORS.md:72]:
```suggestion

```

Is this to be added in this pr? If not, let's not leave TODOs

**Review by lgajowy — COMMENTED** [review]
Some minor fixes. In general the metric adding tutorial looks really good 👍

**Review by lgajowy — APPROVED** [review]

File: CONTRIBUTORS.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (worekleszczy) [inline CONTRIBUTORS.md:72]:
```suggestion
```
